### ISSUE TITLE
[AMD][kernel][Tier 6] Port sol-execbench L2/080_moe_complete_layer_with_shared_expert_backward to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#281]
# FlyDSL implementation: port sol-execbench's `080_moe_complete_layer_with_shared_expert_backward` problem to FlyDSL

## Goal

Port the `080_moe_complete_layer_with_shared_expert_backward` problem from `nvidia/SOL-ExecBench` (subset `L2`) to FlyDSL as part of the sol-execbench → FlyDSL coverage sweep. This is one of 235 problems across 4 subsets (L1, L2, Quant, FlashInfer-Bench) being tracked for FlyDSL kernel coverage.

**Source:** `zai-org/GLM-4.5-Air` (per `definition.json` `hf_id`)

**Description:** Backward pass for complete MoE layer with shared expert. Computes gradients through router (sigmoid + top-k selection), 128 routed experts with SwiGLU, and shared expert with SwiGLU. Gradients flow back through scatter-gather routing pattern.

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — kernel benchmark, not LLM workload)
Subset:    L2
Problem:   080_moe_complete_layer_with_shared_expert_backward
Source:    zai-org/GLM-4.5-Air
```

**Const axes:**
  - `hidden_size` = 4096
  - `moe_intermediate_size` = 1408
  - `n_routed_experts` = 128
  - `num_experts_per_tok` = 8

**Inputs:**
  - `grad_output` [batch_seq_len, hidden_size] bfloat16 — Gradient from next layer
  - `hidden_states` [batch_seq_len, hidden_size] bfloat16 — Original input hidden states (saved from forward)
  - `router_weight` [n_routed_experts, hidden_size] bfloat16 — Router linear projection weights
  - `e_score_correction_bias` [n_routed_experts] float32 — Score correction bias for routing
  - `router_logits` [batch_seq_len, n_routed_experts] float32 — Router logits (saved from forward)
  - `scores` [batch_seq_len, n_routed_experts] float32 — Sigmoid scores (saved from forward)
  - `topk_indices` [batch_seq_len, num_experts_per_tok] int64 — Top-k expert indices per token
  - `topk_weights` [batch_seq_len, num_experts_per_tok] float32 — Top-k routing weights per token
  - `score_mask` [batch_seq_len, n_routed_experts] float32 — Group selection mask
  - `shared_expert_gate_weight` [moe_intermediate_size, hidden_size] bfloat16 — Shared expert gate projection
  - `shared_expert_up_weight` [moe_intermediate_size, hidden_size] bfloat16 — Shared expert up projection
  - `shared_expert_down_weight` [hidden_size, moe_intermediate_size] bfloat16 — Shared expert down projection
  - `shared_gate_output` [batch_seq_len, moe_intermediate_size] bfloat16 — Shared expert gate output (saved from forward)
  - `shared_up_output` [batch_seq_len, moe_intermediate_size] bfloat16 — Shared expert up output (saved from forward)
  - `shared_activated` [batch_seq_len, moe_intermediate_size] bfloat16 — Shared expert activated output (saved from forward)

**Outputs:**
  - `grad_hidden_states` [batch_seq_len, hidden_size] bfloat16 — Gradient w.r.t. input hidden states
  - `grad_router_weight` [n_routed_experts, hidden_size] float32 — Gradient w.r.t. router weights
  - `grad_shared_expert_gate_weight` [moe_intermediate_size, hidden_size] bfloat16 — Gradient w.r.t. shared expert gate weights
  - `grad_shared_expert_up_weight` [moe_intermediate_size, hidden_size] bfloat16 — Gradient w.r.t. shared expert up weights
  - `grad_shared_expert_down_weight` [hidden_size, moe_intermediate_size] bfloat16 — Gradient w.r.t. shared expert down weights

**Workload samples (first 3):**
  - {"uuid": "8686a737-9875-569c-a926-68c9787c7c67", "axes": {"batch_seq_len": 384}, "inputs": {"grad_output": {"type": "custom"}, "hidden_states": {"type": "custom"}, "router_weight": {"type": "custom"},
  - {"uuid": "37876d64-d41a-5f9a-b8be-a72c06e41d56", "axes": {"batch_seq_len": 6144}, "inputs": {"grad_output": {"type": "custom"}, "hidden_states": {"type": "custom"}, "router_weight": {"type": "custom"}
  - {"uuid": "1714936a-08bb-588a-940d-3cca67992d3a", "axes": {"batch_seq_len": 8192}, "inputs": {"grad_output": {"type": "custom"}, "hidden_states": {"type": "custom"}, "router_weight": {"type": "custom"}

```text
Baseline:  sol-execbench PyTorch reference (reference.py) on AMD with torch.compile off
Target:    max_abs_diff < 1e-2 vs PyTorch reference on all tested shapes
           kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
```

## Scope

1. `kernels/sol_execbench/l2_080_moe_complete_layer_with_shared_expert_backward.py` — FlyDSL kernel implementing the op described in `definition.json`, matching the PyTorch reference's numerics
2. `tests/kernels/test_sol_execbench_l2_080_moe_complete_layer_with_shared_expert_backward.py` — parity harness vs PyTorch reference, testing multiple workload shapes
3. `tests/kernels/bench_sol_execbench_l2_080_moe_complete_layer_with_shared_expert_backward.py` — reports `kernel_ms` per shape + comparison vs PyTorch eager baseline
4. `assessment.md` (≤300 words) — ranking which shapes FlyDSL wins/ties/loses vs PyTorch eager

## Acceptance

- Correctness: max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes
- Performance: kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
- Coverage: works on both gfx942 (MI300X) and gfx950 (MI355X)

## Out of scope

- Backward pass (unless the problem is itself a backward op)
- Weight gradients
- Optimizer step
- Cross-arch comparison vs NVIDIA H100 (PyTorch eager on AMD is the baseline)

## Reproduction

Clone `amdpilot-org/SOL-ExecBench`, download the dataset (`scripts/download_data.sh`), find the problem at `data/benchmark/L2/080_moe_complete_layer_with_shared_expert_backward/`, run the sol-execbench driver against a FlyDSL solution. Requires the unified aiter+triton+flydsl Docker image at `/home/xiasun/unified-rocm.Dockerfile` (not yet built).

## Definition of done

- [ ] Kernel implemented at `kernels/sol_execbench/l2_080_moe_complete_layer_with_shared_expert_backward.py`
- [ ] Tests pass on MI300X (gfx942)
- [ ] Tests pass on MI355X (gfx950)
- [ ] Bench shows ≥80% win rate vs PyTorch eager
- [ ] `assessment.md` written

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `080_moe_complete_layer_with_shared_expert_backward` (kernel_id=174). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.054632 ms   (SOL-Score=1.0000, speedup=8.52×)
Scoring Baseline (optimized PyTorch): 0.358976 ms   (SOL-Score=0.4470, speedup=0.86×)
Reference Implementation (plain PyTorch): 1.983712 ms   (SOL-Score=0.1304, speedup=0.19×)

Top public submissions:
  - **Rank 1** (`Hyra`): 0.126919 ms, SOL-Score=0.7836, speedup=2.63× vs reference
  - **Rank 2** (`Recursive`): 0.151349 ms, SOL-Score=0.7157
  - **Rank 3** (`Amir M. Mir | SF Tensor`): 0.162088 ms, SOL-Score=0.7061
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=2.63× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
